### PR TITLE
fix(migration): Decrypt ownCloud secrets v2

### PR DESCRIPTION
**Pull request by ChristophWurst**
* Resolves: # 

## Summary

https://github.com/owncloud/core/pull/38018 and https://github.com/nextcloud/server/pull/20915 are slightly different. This teaches Nextcloud to read encrypted values from a migrated instance.

## TODO

- [x] Do
- [x] Conventional commit (`)`)

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

**Review by miaulalala — APPROVED** [review]
Looks good! :llama:

**Review by AndyScherzinger — APPROVED** [review]
Neighborly 👍

File: lib/private/Security/Crypto.php [hunk 1]
```diff
@@ -114,6 +114,18 @@ private function decryptWithoutSecret(string $authenticatedCiphertext, string $p
 			throw new Exception('Authenticated ciphertext could not be decoded.');
 		}
 
+		// Rearrange arguments for legacy ownCloud migrations
+		$originalParts = $parts;
+		if ($partCount === 4 && $parts[0] === 'v2') {
```
Comment (nickvergessen) [inline lib/private/Security/Crypto.php:119]: *nit* I would save this condition as a variable, so it's clearer that this is the legacy mode

```suggestion
		$legacyOwnCloudMigration = $partCount === 4 && $parts[0] === 'v2'
		if ($legacyOwnCloudMigration) {
```

File: lib/private/Security/Crypto.php [hunk 2]
```diff
@@ -124,7 +136,7 @@ private function decryptWithoutSecret(string $authenticatedCiphertext, string $p
 				$iv = $this->hex2bin($iv);
 			}
 
-			if ($version === '3') {
+			if ($version === '3' || $originalParts[0] === 'v2') {
```
Comment (nickvergessen) [inline lib/private/Security/Crypto.php:139]:
```suggestion
			if ($version === '3' || $legacyOwnCloudMigration) {
```

File: lib/private/Security/Crypto.php [hunk 3]
```diff
@@ -133,8 +145,14 @@ private function decryptWithoutSecret(string $authenticatedCiphertext, string $p
 		$this->cipher->setPassword($encryptionKey);
 		$this->cipher->setIV($iv);
 
-		if (!hash_equals($this->calculateHMAC($parts[0] . $parts[1], $hmacKey), $hmac)) {
-			throw new Exception('HMAC does not match.');
+		if ($partCount === 4 && $originalParts[0] === 'v2') {
```
Comment (nickvergessen) [inline lib/private/Security/Crypto.php:148]:
```suggestion
		if ($legacyOwnCloudMigration) {
```

**Review by nickvergessen — APPROVED** [review]

**Review by Altahrim — APPROVED** [review]

**Pull request merged**